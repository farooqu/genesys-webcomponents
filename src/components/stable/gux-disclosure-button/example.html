<h1>gux-disclosure-button</h1>
<div>
  <h2>Position right</h2>
  <gux-disclosure-button position="right">
    <p slot="panel-content">This panel opens up from the right side.</p>
  </gux-disclosure-button>
</div>
<div>
  <h2>Position left</h2>
  <gux-disclosure-button>
    <p slot="panel-content">This panel opens up from the left side.</p>
  </gux-disclosure-button>
</div>
<div>
  <h2>Position left with is-open</h2>
  <div style="display: flex">
    <div>
      <gux-disclosure-button is-open="true">
        <div style="padding: 0 10px 0 0" slot="panel-content">
          <ul>
            <li>This panel opens up at the start.</li>
            <li>Item 2</li>
            <li>Item 3</li>
          </ul>
        </div>
      </gux-disclosure-button>
    </div>
    <div style="padding: 0 10px">
      <p>Some other container</p>
    </div>
    <div></div>
  </div>
</div>
