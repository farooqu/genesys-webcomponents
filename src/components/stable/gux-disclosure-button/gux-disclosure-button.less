@import (reference) '../../../style/color.less';

:host {
  height: 100%;
  color: @gux-black-50;
}

.gux-disclosure-button-container {
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  height: 100%;

  .gux-disclosure-button {
    width: 16px;
    padding: 0;
    margin: 0;
    color: @gux-black-50;
    background: transparent;
    border-top: none;
    border-right: 1px solid @gux-grey-60;
    border-bottom: none;
    border-left: 1px solid @gux-grey-60;

    gux-icon {
      width: 12px;
      height: 12px;
    }

    &:focus {
      outline: none;
    }
  }

  .gux-disclosure-panel {
    display: none;
    order: -1;
    width: 100%;

    &.gux-active {
      display: block;
    }
  }

  &.gux-right {
    justify-content: flex-end;

    .gux-disclosure-panel {
      order: 1;
    }
  }
}
