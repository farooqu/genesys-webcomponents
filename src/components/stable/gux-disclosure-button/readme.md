# gux-disclosure-button

This is a UI button component that is used to open or dismiss an associated panel.

<!-- Auto Generated Below -->


## Properties

| Property   | Attribute  | Description                                   | Type                | Default     |
| ---------- | ---------- | --------------------------------------------- | ------------------- | ----------- |
| `isOpen`   | `is-open`  | Used to open or close the disclosure panel    | `boolean`           | `false`     |
| `label`    | `label`    | Indicates the label for the disclosure button | `string`            | `undefined` |
| `position` | `position` | Indicates the position of the button panel    | `"left" \| "right"` | `'left'`    |


## Events

| Event    | Description                                              | Type                   |
| -------- | -------------------------------------------------------- | ---------------------- |
| `active` | Return the state of the components panel on state change | `CustomEvent<boolean>` |


## Dependencies

### Depends on

- [gux-icon](../gux-icon)

### Graph
```mermaid
graph TD;
  gux-disclosure-button --> gux-icon
  style gux-disclosure-button fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
