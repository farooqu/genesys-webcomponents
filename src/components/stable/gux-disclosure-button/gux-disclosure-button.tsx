import {
  Component,
  Element,
  Event,
  EventEmitter,
  h,
  JSX,
  Prop,
  State,
  Watch
} from '@stencil/core';

import { randomHTMLId } from '../../../utils/dom/random-html-id';
import { trackComponent } from '../../../usage-tracking';

import { buildI18nForComponent, GetI18nValue } from '../../../i18n';

import translationResources from './i18n/en.json';
import { GuxDisclosureButtonPosition } from './gux-disclosure-button.types';

@Component({
  styleUrl: 'gux-disclosure-button.less',
  tag: 'gux-disclosure-button',
  shadow: true
})
export class GuxDisclosureButton {
  private i18n: GetI18nValue;
  private panelId: string = randomHTMLId('gux-disclosure-button-panel');

  @Element()
  private root: HTMLElement;

  /**
   * Indicates the position of the button panel
   */
  @Prop()
  position: GuxDisclosureButtonPosition = 'left';

  /**
   * Indicates the label for the disclosure button
   */
  @Prop()
  label: string;

  /**
   * Used to open or close the disclosure panel
   */
  @Prop({ mutable: true })
  isOpen: boolean = false;

  /**
   * Indicated image used by button
   */
  @State()
  icon: string = 'arrow-solid-right';

  /**
   * Return the state of the components panel on state change
   * @return the panel state
   */
  @Event()
  active: EventEmitter<boolean>;

  @Watch('isOpen')
  watchIsOpen(): void {
    this.updateIcon();
  }

  changeState(): void {
    this.togglePanel();
    this.active.emit(this.isOpen);
  }

  togglePanel(): void {
    this.isOpen = !this.isOpen;
  }

  updateIcon(): void {
    if (this.position === 'right') {
      this.icon = this.isOpen ? 'arrow-solid-right' : 'arrow-solid-left';
    } else {
      this.icon = this.isOpen ? 'arrow-solid-left' : 'arrow-solid-right';
    }
  }

  async componentWillLoad(): Promise<void> {
    trackComponent(this.root, { variant: this.position });

    this.i18n = await buildI18nForComponent(this.root, translationResources);

    this.updateIcon();
  }

  render(): JSX.Element {
    return (
      <div class={`gux-disclosure-button-container gux-${this.position}`}>
        <button
          class="gux-disclosure-button"
          onClick={() => this.changeState()}
          aria-controls={this.panelId}
          aria-expanded={this.isOpen.toString()}
          aria-label={this.label || this.i18n('defaultLabel')}
        >
          <gux-icon icon-name={`${this.icon}`} decorative></gux-icon>
        </button>
        <div
          id={this.panelId}
          class={{
            'gux-disclosure-panel': true,
            'gux-active': this.isOpen
          }}
          role="region"
        >
          <slot name="panel-content" />
        </div>
      </div>
    ) as JSX.Element;
  }
}
