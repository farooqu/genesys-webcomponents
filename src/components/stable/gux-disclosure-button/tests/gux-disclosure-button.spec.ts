import { newSpecPage } from '@stencil/core/testing';
import { GuxDisclosureButton } from '../gux-disclosure-button';

const components = [GuxDisclosureButton];
const language = 'en';

describe('gux-disclosure-button', () => {
  it('should build', async () => {
    const page = await newSpecPage({
      components,
      html: '<gux-disclosure-button></gux-disclosure-button>',
      language
    });
    expect(page.rootInstance).toBeInstanceOf(GuxDisclosureButton);
  });

  describe('#render', () => {
    [
      '<gux-disclosure-button></gux-disclosure-button>',
      '<gux-disclosure-button is-open></gux-disclosure-button>',
      '<gux-disclosure-button position="left"></gux-disclosure-button>',
      '<gux-disclosure-button is-open position="left"></gux-disclosure-button>',
      '<gux-disclosure-button position="right"></gux-disclosure-button>',
      '<gux-disclosure-button is-open position="right"></gux-disclosure-button>',
      '<gux-disclosure-button label="More Info"></gux-disclosure-button>'
    ].forEach((html, index) => {
      it(`should render component as expected (${index + 1})`, async () => {
        const page = await newSpecPage({ components, html, language });

        expect(page.root).toMatchSnapshot();
      });
    });
  });

  describe('events', () => {
    describe('active', () => {
      it('should open when clicked', async () => {
        const page = await newSpecPage({
          components,
          html: '<gux-disclosure-button></gux-disclosure-button>',
          language
        });
        const element = page.root as HTMLGuxDisclosureButtonElement;
        const button: HTMLButtonElement = element.shadowRoot.querySelector(
          '.gux-disclosure-button'
        );
        const panel = element.shadowRoot.querySelector('.gux-disclosure-panel');
        const icon = element.shadowRoot.querySelector('gux-icon');

        const activeEventSpy = jest.fn();
        element.addEventListener('active', activeEventSpy);

        expect(element.isOpen).toBe(false);
        expect(panel).not.toHaveClass('gux-active');
        expect(icon).toEqualAttribute('icon-name', 'arrow-solid-right');

        button.click();

        await page.waitForChanges();

        expect(element.isOpen).toBe(true);
        expect(panel).toHaveClass('gux-active');
        expect(icon).toEqualAttribute('icon-name', 'arrow-solid-left');
        expect(activeEventSpy).toHaveBeenCalledWith(
          expect.objectContaining({
            detail: true
          })
        );
      });

      it('should close when clicked', async () => {
        const page = await newSpecPage({
          components,
          html: '<gux-disclosure-button is-open></gux-disclosure-button>',
          language
        });
        const element = page.root as HTMLGuxDisclosureButtonElement;
        const button: HTMLButtonElement = element.shadowRoot.querySelector(
          '.gux-disclosure-button'
        );
        const panel = element.shadowRoot.querySelector('.gux-disclosure-panel');
        const icon = element.shadowRoot.querySelector('gux-icon');

        const activeEventSpy = jest.fn();
        element.addEventListener('active', activeEventSpy);

        expect(element.isOpen).toBe(true);
        expect(panel).toHaveClass('gux-active');
        expect(icon).toEqualAttribute('icon-name', 'arrow-solid-left');

        button.click();

        await page.waitForChanges();

        expect(element.isOpen).toBe(false);
        expect(panel).not.toHaveClass('gux-active');
        expect(icon).toEqualAttribute('icon-name', 'arrow-solid-right');
        expect(activeEventSpy).toHaveBeenCalledWith(
          expect.objectContaining({
            detail: false
          })
        );
      });
    });
  });
});
