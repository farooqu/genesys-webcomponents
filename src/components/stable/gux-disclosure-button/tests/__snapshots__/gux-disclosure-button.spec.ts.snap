// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-disclosure-button #render should render component as expected (1) 1`] = `
<gux-disclosure-button>
  <mock:shadow-root>
    <div class="gux-disclosure-button-container gux-left">
      <button aria-controls="gux-disclosure-button-panel-i" aria-expanded="false" aria-label="Disclosure button" class="gux-disclosure-button">
        <gux-icon decorative="" icon-name="arrow-solid-right"></gux-icon>
      </button>
      <div class="gux-disclosure-panel" id="gux-disclosure-button-panel-i" role="region">
        <slot name="panel-content"></slot>
      </div>
    </div>
  </mock:shadow-root>
</gux-disclosure-button>
`;

exports[`gux-disclosure-button #render should render component as expected (2) 1`] = `
<gux-disclosure-button is-open="">
  <mock:shadow-root>
    <div class="gux-disclosure-button-container gux-left">
      <button aria-controls="gux-disclosure-button-panel-i" aria-expanded="true" aria-label="Disclosure button" class="gux-disclosure-button">
        <gux-icon decorative="" icon-name="arrow-solid-left"></gux-icon>
      </button>
      <div class="gux-active gux-disclosure-panel" id="gux-disclosure-button-panel-i" role="region">
        <slot name="panel-content"></slot>
      </div>
    </div>
  </mock:shadow-root>
</gux-disclosure-button>
`;

exports[`gux-disclosure-button #render should render component as expected (3) 1`] = `
<gux-disclosure-button position="left">
  <mock:shadow-root>
    <div class="gux-disclosure-button-container gux-left">
      <button aria-controls="gux-disclosure-button-panel-i" aria-expanded="false" aria-label="Disclosure button" class="gux-disclosure-button">
        <gux-icon decorative="" icon-name="arrow-solid-right"></gux-icon>
      </button>
      <div class="gux-disclosure-panel" id="gux-disclosure-button-panel-i" role="region">
        <slot name="panel-content"></slot>
      </div>
    </div>
  </mock:shadow-root>
</gux-disclosure-button>
`;

exports[`gux-disclosure-button #render should render component as expected (4) 1`] = `
<gux-disclosure-button is-open="" position="left">
  <mock:shadow-root>
    <div class="gux-disclosure-button-container gux-left">
      <button aria-controls="gux-disclosure-button-panel-i" aria-expanded="true" aria-label="Disclosure button" class="gux-disclosure-button">
        <gux-icon decorative="" icon-name="arrow-solid-left"></gux-icon>
      </button>
      <div class="gux-active gux-disclosure-panel" id="gux-disclosure-button-panel-i" role="region">
        <slot name="panel-content"></slot>
      </div>
    </div>
  </mock:shadow-root>
</gux-disclosure-button>
`;

exports[`gux-disclosure-button #render should render component as expected (5) 1`] = `
<gux-disclosure-button position="right">
  <mock:shadow-root>
    <div class="gux-disclosure-button-container gux-right">
      <button aria-controls="gux-disclosure-button-panel-i" aria-expanded="false" aria-label="Disclosure button" class="gux-disclosure-button">
        <gux-icon decorative="" icon-name="arrow-solid-left"></gux-icon>
      </button>
      <div class="gux-disclosure-panel" id="gux-disclosure-button-panel-i" role="region">
        <slot name="panel-content"></slot>
      </div>
    </div>
  </mock:shadow-root>
</gux-disclosure-button>
`;

exports[`gux-disclosure-button #render should render component as expected (6) 1`] = `
<gux-disclosure-button is-open="" position="right">
  <mock:shadow-root>
    <div class="gux-disclosure-button-container gux-right">
      <button aria-controls="gux-disclosure-button-panel-i" aria-expanded="true" aria-label="Disclosure button" class="gux-disclosure-button">
        <gux-icon decorative="" icon-name="arrow-solid-right"></gux-icon>
      </button>
      <div class="gux-active gux-disclosure-panel" id="gux-disclosure-button-panel-i" role="region">
        <slot name="panel-content"></slot>
      </div>
    </div>
  </mock:shadow-root>
</gux-disclosure-button>
`;

exports[`gux-disclosure-button #render should render component as expected (7) 1`] = `
<gux-disclosure-button label="More Info">
  <mock:shadow-root>
    <div class="gux-disclosure-button-container gux-left">
      <button aria-controls="gux-disclosure-button-panel-i" aria-expanded="false" aria-label="More Info" class="gux-disclosure-button">
        <gux-icon decorative="" icon-name="arrow-solid-right"></gux-icon>
      </button>
      <div class="gux-disclosure-panel" id="gux-disclosure-button-panel-i" role="region">
        <slot name="panel-content"></slot>
      </div>
    </div>
  </mock:shadow-root>
</gux-disclosure-button>
`;
